Add Kubernetes on Azure templates

scottslowe opened the pull request:
This PR adds templates for Kubernetes on Azure in Go, Python, TypeScript, and YAML. C# and Java will come in follow-up PRs.

nimbinatus commented on kubernetes-azure-python/__main__.py:
```diff
@@ -0,0 +1,120 @@
+
+# Get some values from the Pulumi configuration (or use defaults)
+azure_cfg = pulumi.Config("azure-native")
+azureLocation = azure_cfg.require("location")
```
any specific reason you're flipping between `require()` and `get()` here?

scottslowe replied:
Yes. The `azure-native:location` configuration value _must_ be present (required by the provider as well as being used in the code), whereas the others have default values supplied if the configuration value isn't present. Is there a better/more recommended way of handling this sort of situation?

nimbinatus replied:
Hm. I honestly don't know; what you're saying makes sense, but I've never used both in the same program, and I don't think the SDK docs make that clear. Maybe adjust the comment on line 9 to something like this?
```python
# Either pull required values from the Pulumi configuration, or get values and use defaults if null
```

scottslowe replied:
I'd overlooked another reason for using both `get()` and `require()` here. The last two project-namespaced configuration values (`mgmtGroupId` and `sshPubKey`) are values where I _can't_ provide a default value; they are too dependent on each customer's environment. I'll update the comments and add another one to better explain.

scottslowe replied:
Comments re-worded and a new comment added in 5b8b6ce

cnunciato replied:
This works, but since we panic when these two required values aren't provided (and there's no indication they're necessary until you run the `pulumi up` after the new-project wizard finishes), do you think it might be better to signal their requirement with config prompts, and just not supply default values?

scottslowe replied:
We can certainly add them to the `template` portion of `Pulumi.yaml` and not supply a default value. For the management group, I think that would probably work fine. However, I suspect the SSH public key will present a problem. You can't use `pulumi config set sshPubKey <key-material-here>` as it will throw an error. You have to use `pulumi config set sshPubKey < <path-to-public-key>`. (It has something to do with multi-line values, AIUI.) Let me try adding them to the `template` portion and see if there are errors with the SSH public key.

scottslowe replied:
Fixed in 270555d

nimbinatus commented on kubernetes-azure-python/__main__.py:
```diff
@@ -0,0 +1,120 @@
+# Get some values from the Pulumi configuration (or use defaults)
+azure_cfg = pulumi.Config("azure-native")
+azureLocation = azure_cfg.require("location")
+proj_cfg = pulumi.Config()
```
Any reason to have a duplicate assignment here with lines 10 and 12? I'm guessing this is a namespacing thing, but I think you probably want a comment explaining that so people don't immediately remove this line and change the `proj_cfg` to `azure_cfg` and break things.

scottslowe replied:
You are correct that it is due to configuration namespacing. I'll add a comment to that effect (and in other languages where a similar mechanism is needed).

scottslowe replied:
Resolved via a comment added in c4dea65

nimbinatus commented on kubernetes-azure-python/__main__.py:
```diff
@@ -0,0 +1,120 @@
+ssh_pub_key = proj_cfg.require("sshPubKey")
+
+# Create an Azure Resource Group
+resource_group = resources.ResourceGroup("resource_group",
```
[PEP8](https://peps.python.org/pep-0008/#indentation) linters will flag all of these instances as not being aligned to PEP8. You have two options:
**Option A**:
```python
resource_group = resources.ResourceGroup("resource_group",
                                         location=azureLocation,
                                         resource_group_name="aks-rg"
                                         )
```

**Option B** (my personal preference because it reduces the line-too-long errors that will happen from other linters):
```python
resource_group = resources.ResourceGroup(
    "resource_group",
    location=azureLocation,
    resource_group_name="aks-rg"
)
```

scottslowe replied:
Switched to Option B in c4dea65, should be resolved now

nimbinatus replied:
Hm, not quite right. I'll make suggestions.

nimbinatus commented on kubernetes-azure-python/__main__.py:
```diff
@@ -0,0 +1,120 @@
+
+# Create an Azure Virtual Network
+virtual_network = network.VirtualNetwork("virtual_network",
+    address_space=network.AddressSpaceArgs(
+        address_prefixes=["10.0.0.0/16"],
+        ),
```
```suggestion
    address_space=network.AddressSpaceArgs(
        address_prefixes=["10.0.0.0/16"],
    ),
```

scottslowe replied:
Fixed in c4dea65

nimbinatus reviewed (commented):
A couple notes on the Python part; I haven't reviewed the others.

nimbinatus commented on kubernetes-azure-python/__main__.py:
```diff
@@ -0,0 +1,122 @@
+ssh_pub_key = proj_cfg.require("sshPubKey")
+
+# Create an Azure Resource Group
+resource_group = resources.ResourceGroup("resource_group",
+    location=azureLocation,
+    resource_group_name="aks-rg"
+)
```
```suggestion
resource_group = resources.ResourceGroup(
    "resource_group",
    location=azureLocation,
    resource_group_name="aks-rg"
)
```

scottslowe replied:
Should be fixed/correct in 5b8b6ce

cnunciato replied:
FWIW, I've been using `black` to format my Python code and it's been working well (I think -- @nimbinatus will let me know if I'm wrong about that!).

scottslowe replied:
I think all these Python formatting errors should be corrected now, thanks!

nimbinatus commented on kubernetes-azure-python/__main__.py:
```diff
@@ -0,0 +1,122 @@
+)
+
+# Create three subnets in the virtual network
+subnet1 = network.Subnet("subnet-1",
+    address_prefix="10.0.0.0/22",
+    resource_group_name=resource_group.name,
+    subnet_name="aks-subnet-1",
+    virtual_network_name=virtual_network.name
+)
```
```suggestion
subnet1 = network.Subnet(
    "subnet-1",
    address_prefix="10.0.0.0/22",
    resource_group_name=resource_group.name,
    subnet_name="aks-subnet-1",
    virtual_network_name=virtual_network.name
)
```

scottslowe replied:
Addressed in 5b8b6ce (should be correct now)

nimbinatus commented on kubernetes-azure-python/__main__.py:
```diff
@@ -0,0 +1,122 @@
+    subnet_name="aks-subnet-1",
+    virtual_network_name=virtual_network.name
+)
+subnet2 = network.Subnet("subnet-2",
+    address_prefix="10.0.4.0/22",
+    resource_group_name=resource_group.name,
+    subnet_name="aks-subnet-2",
+    virtual_network_name=virtual_network.name
+)
```
```suggestion
subnet2 = network.Subnet(
    "subnet-2",
    address_prefix="10.0.4.0/22",
    resource_group_name=resource_group.name,
    subnet_name="aks-subnet-2",
    virtual_network_name=virtual_network.name
)
```

scottslowe replied:
Fixed in 5b8b6ce

nimbinatus commented on kubernetes-azure-python/__main__.py:
```diff
@@ -0,0 +1,122 @@
+    subnet_name="aks-subnet-2",
+    virtual_network_name=virtual_network.name
+)
+subnet3 = network.Subnet("subnet-3",
+    address_prefix="10.0.8.0/22",
+    resource_group_name=resource_group.name,
+    subnet_name="aks-subnet-3",
+    virtual_network_name=virtual_network.name
+)
```
```suggestion
subnet3 = network.Subnet(
    "subnet-3",
    address_prefix="10.0.8.0/22",
    resource_group_name=resource_group.name,
    subnet_name="aks-subnet-3",
    virtual_network_name=virtual_network.name
)
```

nimbinatus commented on kubernetes-azure-python/__main__.py:
```diff
@@ -0,0 +1,122 @@
+)
+
+# Create an Azure Kubernetes Service cluster
+managed_cluster = containerservice.ManagedCluster("managed_cluster",
```
```suggestion
managed_cluster = containerservice.ManagedCluster(
    "managed_cluster",
```

scottslowe replied:
This change is included in 5b8b6ce

nimbinatus reviewed (commented):
Like this for PEP8 for the Python code

scottslowe commented on kubernetes-azure-python/__main__.py:
```diff
@@ -0,0 +1,122 @@
+    location=azureLocation,
+    resource_group_name=resource_group.name,
+    virtual_network_name="aks-vnet"
+)
```
Should be fixed/corrected in 5b8b6ce

cnunciato commented on kubernetes-azure-go/main.go:
```diff
@@ -0,0 +1,183 @@
+			},
+			Location:           pulumi.String(azureLocation),
+			ResourceGroupName:  resourceGroup.Name,
+			VirtualNetworkName: pulumi.String("aks-network"),
```
Similar question here about hard-coding the name.

scottslowe replied:
Fixed via 270555d

scottslowe commented on kubernetes-azure-go/main.go:
```diff
@@ -0,0 +1,183 @@
+
+		// Create an Azure Resource Group
+		resourceGroup, err := resources.NewResourceGroup(ctx, "resourceGroup", &resources.ResourceGroupArgs{
+			Location:          pulumi.String(azureLocation),
+			ResourceGroupName: pulumi.String("aks-rg"),
```
The `azureLocation` variable _is_ pulled from the provider configuration (see line 17).

As for the names, we can certainly remove those lines and let them be auto-generated. I left them there in the event the user wanted to specify a name, but perhaps letting them get auto-generated would be a better approach.

cnunciato replied:
> The azureLocation variable is pulled from the provider configuration (see line 17).

Oh yes -- what I mean is that I _think_ if you _removed_ both line 17 and the references to `Location` throughout, things would "just work", since all Azure Native resources should (by default) use whatever location is configured _on_ the provider.

> As for the names, we can certainly remove those lines and let them be auto-generated.

Ok cool, yeah I'd definitely recommend that then.

scottslowe replied:
Fixed via 270555d

cnunciato commented on kubernetes-azure-go/main.go:
```diff
@@ -0,0 +1,183 @@
+		if err != nil {
+			return err
+		}
+		ctx.Export("resourceGroupName", resourceGroup.Name)
```
This is mostly anecdotal, but I think I've generally seen these clustered up at the end of the program, rather than sprinkled throughout, and this is generally how I've written my programs as well. Do you think we should be consistent there, or do you prefer this approach to that one?

scottslowe replied:
Fixed via 270555d

scottslowe commented:
@cnunciato I've incorporated your feedback on the user experience (adding management group ID and SSH public key to the `template` portion) and I've removed unnecessary location and name settings.

@nimbinatus I believe I've addressed all the formatting issues.

This is ready for another review!

cnunciato commented on kubernetes-azure-go/Pulumi.yaml:
```diff
@@ -0,0 +1,25 @@
+    mgmtGroupId:
+      description: The object ID (from az ad group list) of a group to make cluster admins
+    sshPubKey:
+      description: The public key for SSH access to cluster nodes
```
Maybe a nit, but if I were reading this prompt as a first-time user, my one question would be, "Is this asking me for the content of the key itself or the path to the key file?"

scottslowe replied:
Fixed in 965cc5b (reworded to make clear the contents of the key are needed)

cnunciato commented on kubernetes-azure-go/main.go:
```diff
@@ -0,0 +1,171 @@
+		}
+
+		// Create an Azure Resource Group
+		resourceGroup, err := resources.NewResourceGroup(ctx, "resourceGroup")
```
```suggestion
		resourceGroup, err := resources.NewResourceGroup(ctx, "resourceGroup", &resources.ResourceGroupArgs{})
```

scottslowe replied:
Yep, this caused an error. Fixed in 8cb24a3.

cnunciato commented on kubernetes-azure-go/main.go:
```diff
@@ -0,0 +1,171 @@
+			return err
+		}
+
+		subnet2, err := network.NewSubnet(ctx, "subnet-2", &network.SubnetArgs{
```
When I run this, I get compile errors about this being declared but not used. If we don't need to reference `subnet2` elsewhere in the code, we can ignore it with `_`:

```suggestion
		_, err = network.NewSubnet(ctx, "subnet-2", &network.SubnetArgs{
```

scottslowe replied:
We definitely need subnet1, but aren't referencing subnet2 (or subnet3) in this template. However, their values might be needed if a user adds more nodepools, so I've added them to the list of exports. Fixed in 8cb24a3.

cnunciato commented on kubernetes-azure-go/main.go:
```diff
@@ -0,0 +1,171 @@
+			return err
+		}
+
+		subnet3, err := network.NewSubnet(ctx, "subnet-3", &network.SubnetArgs{
```
Similarly here:

```suggestion
		_, err = network.NewSubnet(ctx, "subnet-3", &network.SubnetArgs{
```

scottslowe replied:
Fixed via 8cb24a3

cnunciato commented on kubernetes-azure-go/main.go:
```diff
@@ -0,0 +1,171 @@
+	pulumi.Run(func(ctx *pulumi.Context) error {
+		// Get some configuration values or set default values
+		cfg := config.New(ctx, "")
+		sshPubKey := cfg.Require("sshPubKey")
+		mgmtGroup := cfg.Require("mgmtGroupId")
```
Noticed there's a note in the Python version that calls these out. Should we do that here and in the other templates also? (I don't have a opinion other than "consistency is nice". 😄)

```suggestion
		sshPubKey := cfg.Require("sshPubKey")
		mgmtGroup := cfg.Require("mgmtGroupId")
```

scottslowe replied:
Fixed in 965cc5b (comment added to all languages)

cnunciato commented on kubernetes-azure-yaml/Pulumi.yaml:
```diff
@@ -0,0 +1,131 @@
+  sshPubKey:
+    type: String
+resources:
+  # Create a new resource group
```
Partly personal preference, but I tend to like a little whitespace between sections and things, just for a little breathing room. Just makes it a little easier to see where certain things end and others begin.

scottslowe replied:
Fixed in 965cc5b (whitespace added to YAML, already present in other languages)

cnunciato reviewed (approved):
Couple things need fixing up, but otherwise just a handful of non-blocking suggestions. Nice work! ⭐
